eth/api_backend.go

```diff
@@ -102,12 +97,8 @@ func (b *EthApiBackend) GetTd(blockHash common.Hash) *big.Int {
 	return b.eth.blockchain.GetTdByHash(blockHash)
 }
 
-func (b *EthApiBackend) GetVMEnv(ctx context.Context, msg core.Message, header *types.Header) (vm.Environment, func() error, error) {
-	stateDb, err := state.New(header.Root, b.eth.chainDb)
-	if err != nil {
-		return nil, nil, err
-	}
-	stateDb = stateDb.Copy()
+func (b *EthApiBackend) GetVMEnv(ctx context.Context, msg core.Message, state ethapi.State, header *types.Header) (vm.Environment, func() error, error) {
+	stateDb := state.(EthApiState).state.Copy()
```

Comment:
Previously it was copied by the caller but now we use ethapi.State interfaces that don't have a Copy function (or any modify function, they're read-only views). So I thought it is easier and maybe also more logical to copy here so that we keep the immutable property of ethapi.State.